src/Middleware/HostFiltering/src/MiddlewareConfigurationManager.cs

```diff
@@ -0,0 +1,120 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.ObjectModel;
+using System.Linq;
+using Microsoft.AspNetCore.Http;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+using Microsoft.Extensions.Primitives;
+
+namespace Microsoft.AspNetCore.HostFiltering;
+
+internal sealed record MiddlewareConfiguration(bool AllowAnyNonEmptyHost, bool AllowEmptyHosts, bool IncludeFailureMessage, ReadOnlyCollection<StringSegment>? AllowedHosts = default);
+
+internal sealed class MiddlewareConfigurationManager
+{
+    private readonly object _syncObject = new();
+    private event Func<HostFilteringOptions>? _getChangeRequestObject;
+    private MiddlewareConfiguration? _middlewareConfiguration;
+    private readonly ILogger<HostFilteringMiddleware> _logger;
+
+    internal MiddlewareConfigurationManager(IOptionsMonitor<HostFilteringOptions> _optionsMonitor, ILogger<HostFilteringMiddleware> logger)
+    {
+        _logger = logger;
+        // Configuration will evaluate during first request
+        _getChangeRequestObject = () => _optionsMonitor.CurrentValue;
+        _optionsMonitor.OnChange(options =>
+        {
+            lock (_syncObject)
+            {
+                _getChangeRequestObject = () => options;
+                if (_logger.IsEnabled(LogLevel.Debug))
+                {
+                    _logger.OptionsChanged(GetLogMessageForHostFilteringOptions(options));
+                }
+            }
+        });
+    }
+    internal MiddlewareConfiguration GetLatestMiddlewareConfiguration()
+    {
+        if (_getChangeRequestObject is not null)
+        {
+            HostFilteringOptions options;
+            lock (_syncObject)
+            {
+                options = _getChangeRequestObject();
+                _getChangeRequestObject = null;
+            }
+            _middlewareConfiguration = ConfigureMiddleware(options);
+        }
+
+        if (_middlewareConfiguration?.AllowAnyNonEmptyHost != true && (_middlewareConfiguration?.AllowedHosts is null || _middlewareConfiguration.AllowedHosts?.Count == 0))
```

Comment:
```suggestion
        var config = _middlewareConfiguration;
        if (!config.AllowAnyNonEmptyHost && (config.AllowedHosts is null || config.AllowedHosts.Count == 0))
```

You can also add this check in the constructor.

Reply:
Why do you want to do all operations in the constructor?
_Constructors should not do much work other than capture the constructor parameters. The cost of any other processing should be delayed until required._

The configuration of the middleware should be evaluated as needed. 
Even from a performance perspective, it's better.
You can have several HostFilteringOptions updates (e.g. 5 ) but only one request after those updates so reconfiguration should be done only once.
That's why I created the delegate, to transfer responsibility from the constructor to the method when configuration is needed.

Reply:
Constructors should make sure the object is initialized and ready to use. This data is going to be needed on all requests, so it makes sense to get it ready in advance. It's not a lot of extra work in this case.

> You can have several HostFilteringOptions updates (e.g. 5 ) but only one request after those updates so reconfiguration should be done only once.

Requests are _much_ more common than options updates.

Reply:
> Constructors should make sure the object is initialized and ready to use. This data is going to be needed on all requests, so it makes sense to get it ready in advance. It's not a lot of extra work in this case.

why my implementation is wrong? it is doing the same as you mentioned it is initializing data to be ready to use, with the difference that middleware initialization is faster and reevaluation is done only when needed, you can have a bigger list of allowed hosts that will use not needed resources to reevaluate state.

> Requests are _much_ more common than options updates.
You are right but it is theoretical

Reply:
My main complaint is that the implementation is more complex than necessary. Keep it simple.

As for perf, avoid doing work on a request thread when possible, do it at startup or on an event thread instead.

Reply:
I can partially agree as it is better to do work on an event thread :D but about complexity, I think it is still simple 10 extra lines with logging. 
We have two different concepts which have their own pros and cons.
I kept the current implementation because the configuration was evaluated during the request, I also agree with this because it is better to avoid unnecessary evaluation. Re-evaluation on the event thread is slowing down event processing, any nested methods assigned to an event will be delayed. 

You maintain this code so I will change it as you requested :)